### PR TITLE
feat: Add column cleanup to reduce memory usage in flows

### PR DESCRIPTION
**Pull request by RohanAwhad**
## Summary
- Add `output_columns` field to FlowMetadata for specifying required output columns
- Add `columns_to_keep` param to `flow.generate()` for runtime overrides
- Add `all_output_columns` property to inspect available columns
- Add `get_column_summary()` method to list input/intermediate/output columns
- Validation: `columns_to_keep` must be superset of `output_columns` (lower bound)
- Validation: `columns_to_keep` must be subset of available columns (upper bound)
- Phase 1: Drop intermediate columns after flow completion
- Phase 3: Early column dropping during execution (memory optimization)
- Original input columns are auto-preserved

## Test plan
- [x] `output_columns` field works correctly
- [x] `all_output_columns` property returns all block outputs
- [x] `get_column_summary()` returns correct input/intermediate/output columns
- [x] `columns_to_keep` param keeps additional columns beyond output_columns
- [x] Lower bound validation: missing output columns raises error
- [x] Upper bound validation: unknown columns raises error
- [x] `ColumnDependencyTracker` correctly identifies droppable columns
- [x] All 31 column cleanup unit tests pass

## Test script

```python
# Test script for column cleanup feature
import pandas as pd
from sdg_hub.core.flow.base import Flow
from sdg_hub.core.flow.metadata import DatasetRequirements, FlowMetadata
from sdg_hub.core.blocks.transform.text_concat import TextConcatBlock
from sdg_hub.core.blocks.transform.duplicate_columns import DuplicateColumnsBlock

# Create a flow with output_columns and dataset_requirements
flow = Flow(
    metadata=FlowMetadata(
        name="test",
        output_columns=["final"],
        optimize_memory=True,
        dataset_requirements=DatasetRequirements(
            required_columns=["a", "b"],
        ),
    ),
    blocks=[
        TextConcatBlock(
            block_name="step1",
            input_cols=["a", "b"],
            output_cols="intermediate",
        ),
        DuplicateColumnsBlock(
            block_name="step2",
            input_cols={"intermediate": "final"},
        ),
    ],
)

# Get column summary
print("Column Summary:")
summary = flow.get_column_summary()
print(f"  Input columns: {summary['input']}")
print(f"  Intermediate columns (will be dropped): {summary['intermediate']}")
print(f"  Output columns: {summary['output']}")

# Run the flow
print("\nRunning flow...")
dataset = pd.DataFrame({"a": ["hello"], "b": ["world"]})
result = flow.generate(dataset)
print(f"\nFinal output columns: {list(result.columns)}")
```

## Test output

```
Column Summary:
  Input columns: ['a', 'b']
  Intermediate columns (will be dropped): ['intermediate']
  Output columns: ['final']

Running flow...
[16:17:47] INFO     Starting flow 'test' v1.0.0 with 1 samples       base.py:583
                    across 2 blocks                                             
           INFO     Memory optimization enabled - will drop unused   base.py:600
                    columns                                                     
           INFO     Executing block 1/2: step1 (TextConcatBlock)     base.py:756
╭─────────────────────────────────── step1 ────────────────────────────────────╮
│ 📊 Processing Input Data                                                     │
│ Block Type: TextConcatBlock                                                  │
│ Input Rows: 1                                                                │
│ Input Columns: 2                                                             │
│ Column Names: a, b                                                           │
│ Expected Output Columns: intermediate                                        │
╰──────────────────────────────────────────────────────────────────────────────╯
╭────────────────────────────── step1 - Complete ──────────────────────────────╮
│ ✅ Processing Complete                                                       │
│ Rows: 1 → 1                                                                  │
│ Columns: 2 → 3                                                               │
│ 🟢 Added: intermediate                                                       │
│ 📋 Final Columns: a, b, intermediate                                         │
╰──────────────────────────────────────────────────────────────────────────────╯
           INFO     Block 'step1' completed successfully: 1 samples, base.py:804
                    3 columns                                                   
           INFO     Executing block 2/2: step2                       base.py:756
                    (DuplicateColumnsBlock)                                     
╭─────────────────────────────────── step2 ────────────────────────────────────╮
│ 📊 Processing Input Data                                                     │
│ Block Type: DuplicateColumnsBlock                                            │
│ Input Rows: 1                                                                │
│ Input Columns: 3                                                             │
│ Column Names: a, b, intermediate                                             │
│ Expected Output Columns: final                                               │
╰──────────────────────────────────────────────────────────────────────────────╯
╭────────────────────────────── step2 - Complete ──────────────────────────────╮
│ ✅ Processing Complete                                                       │
│ Rows: 1 → 1                                                                  │
│ Columns: 3 → 4                                                               │
│ 🟢 Added: final                                                              │
│ 📋 Final Columns: a, b, final, intermediate                                  │
╰──────────────────────────────────────────────────────────────────────────────╯
           INFO     Block 'step2' completed successfully: 1 samples, base.py:804
                    4 columns                                                   
           INFO     Dropping 1 unused columns: ['intermediate']      base.py:816

╭────────────────────────────── test - Complete ───────────────────────────────╮
│                            Flow Execution Summary                            │
│ ┏━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━┳━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━━━━┳━━━━━━━┓ │
│ ┃ Block Name       ┃ Type        ┃ Durat… ┃   Rows   ┃   Columns   ┃ Stat… ┃ │
│ ┡━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━╇━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━━━━╇━━━━━━━┩ │
│ │ step1            │ TextConcat… │  0.00s │  1 → 1   │     +1      │   ✓   │ │
│ │ step2            │ DuplicateC… │  0.00s │  1 → 1   │     +1      │   ✓   │ │
│ ├──────────────────┼─────────────┼────────┼──────────┼─────────────┼───────┤ │
│ │ TOTAL            │ 2 blocks    │  0.00s │ 1 final  │   3 final   │  2/2  │ │
│ └──────────────────┴─────────────┴────────┴──────────┴─────────────┴───────┘ │
╰──────────────────────────────────────────────────────────────────────────────╯

           INFO     Flow 'test' completed successfully: 1 final      base.py:706
                    samples, 3 final columns                                    

Final output columns: ['a', 'b', 'final']
```

## Unit tests

```
$ pytest tests/flow/test_column_cleanup.py -v
31 passed
```

## Release Notes

* **New Features**
  * Added memory optimization for data pipelines with ability to specify output columns to retain
  * Added column dependency tracking to automatically drop unused intermediate columns during execution
  * Added flow metadata configuration options for output columns and memory optimization

* **Enhancements**
  * Extended Flow API with column inspection capabilities to view all outputs and categorize column usage
  * Updated flow execution and generation methods to support selective column retention

* **Tests**
  * Added comprehensive test coverage for column cleanup and memory optimization workflows

### CODE REVIEW

**Review by eshwarprasadS — DISMISSED** [review]
Thanks for the addition @RohanAwhad

I see that we already have an `output_columns` field in flow yamls that is populated by flow authors. Can we either co-opt it or replace it with the new incoming `final_output_columns` field, in order to not cause confusion or deadcode?

Also, I think as we discussed before, it would be good to also add provisions for 2 kinds of users: 

1. Flow Authors: For flow authors, the currently designed `final_output_columns` can be used, but I suggest naming it something along the lines of `minimal_output_columns` (Referring to the fact that these columns are required at a minimum for this flow to be useful). 

2. Flow users / developers: For consumers of the flow, I think we might want to add capabilities through`flow.generate()` api with an additional parameter, that can accept the `columns_to_keep`. As we discussed before, this `columns_to_keep` can be a set thats lower bounded by the `minimal_output_columns` and upper bounded by the superset of all available output columns of the flow (this should be calculable by us programmatically with the information from each of the constituent blocks). 

I think an option to access `all_output_columns` of a flow would also be good to have for the user experience (a user can know what columns are available to choose from). 

Some critical validation here can be to make sure that the `columns_to_keep` list is a superset of `minimal_output_columns` and subset of `all_output_columns`

File: src/sdg_hub/core/flow/base.py [hunk 1]
```diff
@@ -498,6 +559,17 @@ def generate(
...
+        # Initialize column tracker for memory optimization
+        if self.metadata.optimize_memory and effective_columns_to_keep:
+            self._column_tracker = ColumnDependencyTracker(
+                self.blocks,
+                effective_columns_to_keep,
+                original_columns,
+            )
+            flow_logger.info("Memory optimization enabled - will drop unused columns")
+        else:
+            self._column_tracker = None
```
[the remark this answers is not in the document]
Comment (RohanAwhad) [reply src/sdg_hub/core/flow/base.py]: nope it should not. by default it should mean that keep all columns

**Comment by RohanAwhad** [comment]
@shivchander Renamed `minimal_output_columns` → `output_columns` as discussed. Removed the backwards compat migration since we're using the original field name directly now.

File: src/sdg_hub/core/flow/execution.py [hunk 2]
```diff
@@ -359,6 +415,14 @@ def execute_flow(
     # Convert to DataFrame if needed (backwards compatibility)
     dataset, was_dataset = convert_to_dataframe(dataset)
 
+    # Capture original columns for preservation during cleanup
+    original_columns = set(dataset.columns)
+
+    # Determine effective output columns from metadata
+    effective_output_columns = (
+        set(flow.metadata.output_columns) if flow.metadata.output_columns else None
```
[the remark this answers is not in the document]
Comment (eshwarprasadS) [reply src/sdg_hub/core/flow/execution.py:433]: The truthy check is intentional here. We've added explicit validation in `FlowMetadata` that rejects `output_columns=[]` with a clear error message: "output_columns must not be empty when specified. Remove the field entirely to keep all columns."

So the three states are:

- `output_columns` *not specified* (None) → keep all columns, no cleanup
- `output_columns=[]` → *validation error* at flow construction time
- `output_columns=["col1", "col2"]` → cleanup enabled, keep only these + original input columns
Since `[]` can never reach the execution path, the truthy check in `execute_flow()` is equivalent to an `is not None` check — both correctly distinguish "not configured" from "configured with columns."

File: src/sdg_hub/core/flow/execution.py [hunk 3]
```diff
@@ -455,6 +519,17 @@ def execute_flow(
     flow._block_metrics = []
     run_start = time.perf_counter()
 
+    # Initialize column tracker for early dropping of unused columns
+    if effective_output_columns:
+        flow._column_tracker = ColumnDependencyTracker(
```
[the remark this answers is not in the document]
Comment (eshwarprasadS) [reply src/sdg_hub/core/flow/execution.py]: The tracker is built from static `block.input_cols` intentionally. In practice, `runtime_params` are used for execution parameters like `temperature`, `max_tokens`, or prompt configs — not for structural fields like input_cols/output_cols. No existing flow uses `runtime_overrides` for column specifications, and overriding `input_cols` at runtime would be an unusual pattern.

That said, if a user did override `input_cols` via `runtime_params`, the block would temporarily read different columns than what the tracker expects. In the worst case, the tracker might drop a column that the overridden block needs, which would surface immediately as a `MissingColumnError` from the block's own validation (`BaseBlock._validate_columns`). So the failure mode is a clear error, not silent data corruption.

If this becomes a real use case in the future, we can rebuild the tracker after applying runtime overrides. For now, the static analysis covers all existing flows correctly.

**Comment by shivchander** [comment]
# PR Review Summary: Column Cleanup Feature

## Critical Issues (1 found)

- **Column drop error misattributed as block failure** (`execution.py:273-283`): The column-dropping code runs inside the broad `except Exception` block that catches block execution errors. If `current_dataset.drop(columns=cols_to_drop)` fails, the error is reported as `"Block 'X' execution failed"` — when the block actually succeeded. **Fix:** Move column-drop logic outside the block try/except or add a dedicated try/except with a distinct error message.

## Important Issues (6 found)

1. **Missing output columns only warns, silent data loss** (`execution.py:344-347`): When a user declares `output_columns` and they're absent from the final dataset, only a warning is logged. This is a contract violation — should raise `FlowValidationError` or at least log at `error` level.

2. **Silent empty-list fallback for unexpected types** (`column_tracker.py:72`): `_extract_input_columns` returns `[]` for unexpected types instead of raising, unlike `BaseBlock._normalize_columns()` which raises `ValueError`. Could cause premature column dropping.

3. **No early validation of output_columns** (`metadata.py` + `execution.py`): No check that `output_columns` references columns the flow's blocks actually produce. Users only discover misconfiguration after expensive flow execution completes.

4. **Poor encapsulation on `ColumnDependencyTracker`**: All attributes (`blocks`, `columns_to_keep`, `original_columns`, `last_consumer`) are public and mutable. External mutation can silently corrupt tracker logic. Use `frozenset` for preservation sets; don't store `blocks` (only needed during construction).

5. **`_column_tracker` breaks `PrivateAttr` encapsulation** (`base.py:57`): Stored as `PrivateAttr` on `Flow` but accessed directly by free function `execute_blocks_on_dataset` via `flow._column_tracker`. Should be passed as an explicit parameter instead.

6. **`FlowMetadata` docstring missing `output_columns`** (`metadata.py:206-228`): The `Attributes` section omits the newly added field.

## Test Coverage Gaps (4 found)

1. **Early dropping not actually verified** (Criticality: 8/10): `test_flow_drops_columns_early_during_execution` only checks final output — `_cleanup_final_columns` would produce the same result even if early dropping is broken. Need to capture columns present at each block's execution time.

2. **No direct `_cleanup_final_columns` unit test** (Criticality: 8/10): Edge cases (empty drop set, all output columns missing) are untested.

3. **No checkpointing + column cleanup tests** (Criticality: 8/10): Schema mismatches between checkpoint data and new data with different column cleanup configs could cause issues.

4. **Missing edge case tests** (Criticality: 5-7/10): Non-consecutive block column reuse, whitespace-only `output_columns`, overlapping output/original columns, single-block flow.

## Suggestions

- Docstring says "Build dependency graph" but method builds a flat last-consumer map (`column_tracker.py:42`)
- `ColumnDependencyTracker` class docstring should mention preservation semantics
- `execute_flow` docstring doesn't mention column cleanup behavior
- `_extract_input_columns` duplicates logic from `BaseBlock._normalize_columns` — risk of drift
- Silent filtering of non-string entries in `validate_output_columns` should reject rather than drop
- Clean up `_column_tracker` in `finally` block after execution to avoid stale state
- `dry_run` doesn't simulate column cleanup — should document this limitation
- Import ordering in test file — `# Third Party` / `# First Party` comments misplaced

## Strengths

- Clean, focused single-responsibility design for `ColumnDependencyTracker`
- Fully backward compatible — feature only activates when `output_columns` is specified
- Comprehensive test coverage of core happy paths with real blocks (not mocks)
- Excellent user-facing error messages (e.g., "Remove the field entirely to keep all columns")
- Original input columns always preserved — good UX decision
- Appropriate logging at each cleanup step

## Recommended Action

1. Fix the critical issue — move column-drop logic outside the block execution try/except
2. Address the 3 high-severity error handling issues
3. Improve encapsulation — `frozenset`, remove stored `blocks`, pass tracker as parameter
4. Strengthen tests — verify early dropping mid-execution, add `_cleanup_final_columns` tests
5. Update docstrings — add `output_columns` to `FlowMetadata`, fix `_build_dependency_graph`

---
🤖 Generated with [Claude Code](https://claude.com/claude-code)

**Review by shivchander — APPROVED** [review]
